src/main/grammar/SourcePawn_test.bnf

```diff
@@ -0,0 +1,197 @@
+    NEW_KEYWORD="new"
+    OPERATOR_KEYWORD="operator"
+    PUBLIC_KEYWORD="public"
+    STATIC_KEYWORD="static"
+    STOCK_KEYWORD="stock"
+
+    DECL_KEYWORD="decl"
+    NEW_KEYWORD="new"
+    PUBLIC_KEYWORD="public"
+
+    // SourcePawn-specific keywords
+    ENUM_KEYWORD="enum"
+    STRUCT_KEYWORD="struct"
+
+    // TODO: remove me (for live preview purposes only)
+    SPACE="regexp:[ \n\r\t\f]"
+
+    line_comment="regexp://[^\r\n]*"
+    block_comment="regexp:/\*(.|\n)*?\*/"
+    preprocessor_comment="regexp:#(assert|define|else|elseif|endif|endinput|error|file|if|include|line|pragma|section|tryinclude|undef)[^\r\n]*"
+
+    integer_literal="regexp:[-+]?\d[_\d]*"
+    float_literal="regexp:[-+]?\d[_\d]*\.\d[_\d]*(e[-+]?\d+)?"
+    hex_literal="regexp:[-+]?0x[a-fA-F0-9_]+"
+    binary_literal="regexp:[-+]?0b[01_]+"
+
+    string_literal="regexp:\"(\\.|[^\"])*\""
+    character_literal="regexp:'(\\.|[^\"])'"
+
+    // @ and _ are not themselves valid identifiers.
+    identifier="regexp:([@_a-zA-Z][@_a-zA-Z0-9]+|[a-zA-Z][@_a-zA-Z0-9]*)"
+
+
+    // xxx///////////////
+    // Below is from an attempt to use dvander's grammar as a base
+    // See https://wiki.alliedmods.net/SourcePawn_Transitional_Syntax#Grammar
+
+    // This is dvander's symbol. It doesn't handle all the edge cases Pawn allows
+//    symbol="regexp:[A-Za-z_]([A-Za-z0-9_@]*)"
+    symbol="regexp:([@_a-zA-Z][@_a-zA-Z0-9]+|[a-zA-Z][@_a-zA-Z0-9]*)"
+  ]
+}
+
+
+root ::= root_item *
+private root_item ::= (func | global_assn | global) ';'
+
+global_assn ::= global '=' literal
+
+global ::= global_new | global_old
+private global_new ::= storage_class* type_expr symbol old_dims?
+private global_old ::= storage_class* tag? symbol old_dims?
+
+func ::= global func_arglist func_body
+func_arglist ::= '(' (argdecl (',' argdecl)*)? ')'
+func_body ::= '{' (stmt ';')* '}'
+
+stmt ::= vardecl | local_assn | method_call
+local_assn ::= (vardecl | symbol) '=' value_expr
+
+value_expr ::= simple_value_expr
+simple_value_expr ::= literal | method_call | symbol | '(' value_expr ')'
+
+cmp_expr
+private cmp_op ::=
+
+method_call ::= (func_qn | SIZEOF_KEYWORD) '(' (value_expr (',' value_expr)*)? ')' {pin=2}
+func_qn ::= symbol ('.' symbol)?
+
+if_stmt ::= IF_KEYWORD '('
+
+
+return_type ::= return_new | return_old
+return_new ::= type_expr new_dims?        // Note, dims not yet supported.
+return_old ::= old_dims? label?
+
+argdecl ::= arg_new | arg_old
+arg_new ::= "const"? type_expr '&'? symbol old_dims? ('=' arg_init)?
+arg_old ::= "const"? tags? '&'? symbol old_dims? ('=' arg_init)?
+arg_init ::= literal | symbol
+
+vardecl ::= var_old | var_new
+var_new ::= var_new_prefix? type_expr symbol old_dims?
+var_new_prefix ::= "static" | "const"
+var_old ::= var_old_prefix tag? symbol old_dims?
+var_old_prefix ::= "new" | "decl" | "static" | "const"
+
+storage_class ::= "public" | "static" | "const" | "stock"
+
+type_expr ::= (builtin_type | symbol) new_dims?
+builtin_type ::= "void"
```

Comment:
Here is missing the `any` type. Also I'm not sure how the `Handle` type should be handled, sourcepawn declares it as an `enum` but it is actually treated as a type.

Reply:
Ah, word. I skipped doing tags and some other pre-1.7 stuff — I presume `any` is only used as a tag (i.e. `new any:myvar = 123`) and not a fancy new type (e.g. `any myvar = 123;`). Lemme add all that in.

Thanks for taking a look here :)

Reply:
Yeah, `any` accepts any type, here are some valid examples:
```
    any a = 'a';
    any b = 1;
    any c = 1.1;
    any d = new ArrayList(); // Or any other handle
    any e[] = "hello"; // `any x = ""` is not valid


 ```

Reply:
Oh, shit, word – like a C++ `auto`. I'll add her in.

I'll push up my latest changes later tonight. I believe I've since added:
 - Dimensions for declarations
 - The `view_as` operator
 - methodmaps, including properties (though, I missed inheritance and destructors)
 - native & forward declarations
 - I think I added function calls

Still missing:
 - typedefs
 - typesets
 - enum structs
 - new/delete expressions
 - Support for function calls with named parameters

Reply:
The `sizeof` operator is missing as well, it should be noted that both these annotations are valid:
```
sizeof(var); // Format(var, sizeof(var), ...);
sizeof var // Format(var, sizeof var, ...);
```

Also the `any` type iirc is not like C++ `auto` since you can reassign the variable to any type you want:
```
any x = 1;
x = 2.1;
x = '';
x = new ArrayList();
```